### PR TITLE
use aws web identity role provider

### PR DESCRIPTION
**Pull request by tennix**
#### What is the purpose of this change?

Support EKS iam role.

#### Was the change discussed in an issue or in the forum before?

#3753 

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [ ] I have added tests for all changes in this PR if appropriate.
- [ ] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

I've tested this manually in a kops cluster with kube2iam deployed.

### CODE REVIEW

**Comment by ncw** [comment]
That looks very useful - thanks :-)

Can you update the auth docs here too please: https://github.com/rclone/rclone/blob/master/docs/content/s3.md#authentication

(All the other docs are auto generated from that one file)

**Comment by tennix** [comment]
@ncw I've updated the document. The CI fails, but I don't think this is related to my changes :slightly_smiling_face:

**Comment by ncw** [comment]
That looks great now - thanks!

Yes the CI is broken for some unknown reason :-(

File: backend/s3/s3.go [hunk 1]
```diff
@@ -985,6 +986,11 @@ func s3Connection(opt *Options) (*s3.S3, *session.Session, error) {
 			}),
 			ExpiryWindow: 3 * time.Minute,
 		},
+
+		// Pick up IAM role if we are in EKS
+		&stscreds.WebIdentityRoleProvider{
```
Comment (tirsen) [inline backend/s3/s3.go]: Does creating this struct this way really work? It has quite a lot of non-public fields that are set when calling `NewWebIdentityRoleProvider`. As you can see that state is being used in the `Retrieve` function: https://github.com/aws/aws-sdk-go/blob/master/aws/credentials/stscreds/web_identity_provider.go#L67

Comment (ncw) [reply backend/s3/s3.go]: @tirsen - is this causing you a problem?

Want to send a PR to fix?

Comment (tirsen) [reply backend/s3/s3.go]: Yeah EKS role authentication doesn't work. I've checked all the env vars of the pod but everything seems to be in good order. Tracing through the code paths of `WebIdentityRoleProvider` I can see that the env vars will not be read if that struct is created that way.

Here is a fix:
https://github.com/rclone/rclone/pull/4023

**Pull request merged**